### PR TITLE
subsys: random: device id seed for Xoroshiro128+ in test mode

### PR DESCRIPTION
**Pull request by alexanderwachter**
Adds the possibility to seed Xoroshiro128+ with the hwinfo device id.
This is only allowed in random test mode since the device id is
not a source of entropy and will produce the same series of numbers
after every boot.

### CODE REVIEW

**Comment by aescolar** [comment]
Just my 2 cents (I do not maintain in any way this area): It seems to me the kconfig options of the entropy and random drivers, regarding the TEST_RANDOM_GENERATOR, are a bit of a moebius loop and need some re-work.
Regarding the new option and how you add the file in the Cmakelists, it seems to be a bit confusing: You are just changing how the xoroshiro is seeded, but it is otherwise the same. If anything I would have added a new option like XOROSHIRO_SEED_FROM_DEV_ID (But I can see that you add it this way due to the incompatibility of ENTROPY_HAS_DRIVER and TEST_RANDOM_GENERATOR)

**Comment by aescolar** [comment]
I think that it would be nice if who added TEST_RANDOM_GENERATOR, could explain what was the motivation for it depending on !ENTROPY_HAS_DRIVER.
I fail to see the motivation between this Kconfig options, and therefore is a bit difficult for me to understand how they should be used.

File: subsys/random/rand32_xoroshiro128.c [hunk 1]
```diff
@@ -76,6 +78,32 @@ static int xoroshiro128_initialize(struct device *dev)
...
+#elif CONFIG_XOROSHIRO_DEV_ID_RANDOM_GENERATOR
+static u32_t xoroshiro128_next(void);
+
+static int xoroshiro128_initialize(struct device *dev)
+{
+	ssize_t id_len;
+	int i;
+
+	id_len = hwinfo_get_device_id((u8_t *)&state, sizeof(state));
+	if (id_len < 1) {
```
Comment (lpereira) [inline subsys/random/rand32_xoroshiro128.c]: Is the check here correct, `id_len < 1`?

File: subsys/random/Kconfig [hunk 2]
```diff
@@ -52,4 +52,14 @@ config XOROSHIRO_RANDOM_GENERATOR
...
 	  It is so named because it uses 128 bits of state.
 
+config XOROSHIRO_DEV_ID_RANDOM_GENERATOR
+	bool "Use Xoroshiro128+ without entropy as PRNG"
+	depends on (DEVICE_ID_HAS_DRIVER && TEST_RANDOM_GENERATOR)
+	help
+	  Enables the Xoroshiro128+ pseudo-random number generator, that
+	  uses the hwinfo device id driver as a seed source. This is not a
+	  cryptographically secure random number generator and it produces a
+	  predictable series of "random" numbers that is the same after every
```
Comment (lpereira) [inline subsys/random/Kconfig]: I was going to say that I was against this patch until I saw that `CONFIG_TEST_RANDOM_GENERATOR` is required for this to work.  I'm fine with it this way.

Maybe just clarify that two different devices will generate distinct numbers?

File: subsys/random/rand32_xoroshiro128.c (hunk 1 above)
Comment (alexanderwachter) [reply subsys/random/rand32_xoroshiro128.c]: Negative number means the id is not available and zero would give no random numbers at all. With one we get at least somehow a random number.

File: subsys/random/Kconfig (hunk 2 above)
Comment (alexanderwachter) [reply subsys/random/Kconfig]: We can't guarantee that two different devices ALWAYS have different IDs. That's why we renamed it from get unique id to get id. But the probability is very high that they differ from device to device.

**Comment by lpereira** [comment]
> I think that it would be nice if who added TEST_RANDOM_GENERATOR, could explain what was the motivation for it depending on !ENTROPY_HAS_DRIVER.

It's like this for two reasons:

1. To facilitate writing new ports.  It's not a driver you absolutely must write if all you want while porting Zephyr to a new SoC is to run something like the `philosophers` sample. Once you get that right, you can implement an entropy driver for that SoC, so that you can have, say, a TCP stack on your system.
2.  The test random generators are there to run tests on emulators (e.g. sanitycheck).  They're not there to be used on real devices.  Random number generators are necessary to test stuff like the network layer (e.g. TCP sequence numbers), test stack canaries, etc.  Qemu supports a random device (uses the host `/dev/random` for instance), and if drivers are written, the need for `CONFIG_TEST_RANDOM_GENERATOR` are greatly reduced.

**Comment by aescolar** [comment]
Thanks @lpereira,
Then, shouldn't TEST_RANDOM_GENERATOR be ```default y if !ENTROPY_HAS_DRIVER```, instead of ```depends on !ENTROPY_HAS_DRIVER```
Meaning, if there is no entropy source, by default select the test one, but even if it were to be an entropy driver, do not prevent people from selecting the tets one (if they want to, say, have reproducible sequences for testing)

**Comment by lpereira** [comment]
I don't like the idea of defaulting these generators to `y` if there's no entropy driver. It makes it easy to overlook them, and given that they're used for a (admittedly weak) security feature, it's best when users are aware of the implications. They're really just for testing purposes. In an ideal world, they would be removed altogether.

**Comment by aescolar** [comment]
@lpereira : Fair enough, so should they then be left to default to n, but not depend on there not being an entropy driver?

**Comment by alexanderwachter** [comment]
@lpereira can you give your +1?

**Pull request closed without merging**